trade-finance-manager-api/src/v1/controllers/auth/auth-provider.js

```diff
@@ -0,0 +1,207 @@
+   * @param {String} csrfToken: CSRF token
+   * @param {Object} authCodeUrlRequestParams: Parameters for requesting an auth code url
+   * @param {Object} authCodeRequestParams: Parameters for requesting tokens using auth code
+   * @param {Object} msalInstance: MSAL instance
+   * @return {Object} PKCE codes, auth code request, login URL.
+   */
+  async getAuthCodeUrl(csrfToken, authCodeUrlRequestParams, authCodeRequestParams, msalInstance) {
+    /**
+     * Generate Proof Key of Code Exchange (PKCE) Codes,
+     * before starting the authorization flow.
+     */
+    const { verifier, challenge } = await this.cryptoProvider.generatePkceCodes();
+
+    // Set generated PKCE codes and method as session vars
+    const response = { csrfToken };
+
+    response.pkceCodes = {
+      challengeMethod: 'S256',
+      verifier,
+      challenge,
+    };
+
+    /**
+     * By manipulating the request objects below before each request, we can obtain
+     * auth artifacts with desired claims. For more information, visit:
+     * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationurlrequest
+     * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationcoderequest
+     * */
+    response.authCodeUrlRequest = {
+      ...authCodeUrlRequestParams,
+      redirectUri: this.config.redirectUri,
+      responseMode: 'form_post', // recommended for confidential clients
+      codeChallenge: response.pkceCodes.challenge,
+      codeChallengeMethod: response.pkceCodes.challengeMethod,
+    };
+
+    response.authCodeRequest = {
+      ...authCodeRequestParams,
+      redirectUri: this.config.redirectUri,
+      code: '',
+    };
+
+    response.loginUrl = await msalInstance.getAuthCodeUrl(response.authCodeUrlRequest);
+
+    return response;
+  }
+
+  /**
+   * handleRedirect
+   * Modified Microsoft example, to meet out minimal needs.
+   * OpendId tokenResponse is enough for us (UKEF), it has all claims (aka fields) required for user data.
+   * We don't save Authorisation code because we don't need Access token.
+   * @param {Object} pkceCode: PKCE Code object
+   * @param {Object} origAuthCodeRequest: Original auth code request
+   * @param {String} code: authZ code
+   * @param {Object} req: Request object
+   */
+  // TODO: "returns" documentation
+  // TODO: update body param to consume required property in the body,
+  // instead of passing around req.body
+  async handleRedirect(pkceCode, origAuthCodeRequest, code, body) {
+    const authCodeRequest = {
+      ...origAuthCodeRequest,
+      code, // authZ code
+      codeVerifier: pkceCode.verifier, // PKCE Code Verifier
+    };
+
+    const msalInstance = this.getMsalInstance(this.config.msalConfig);
+
+    // TODO: do we need token cache?
+    // msalInstance.getTokenCache().deserialize(req.session.tokenCache);
```

Comment:
I don't think we need this because we handle our own tokens, sessions, JWT etc. Unless we want to deviate away from this (I don't recommend), we can remove these comments.